gusto/advection.py

```diff
@@ -1,34 +1,97 @@
 from __future__ import absolute_import
 from abc import ABCMeta, abstractmethod
-from firedrake import Function, TestFunction, TrialFunction, \
-    LinearVariationalProblem, LinearVariationalSolver, FacetNormal, \
-    dx, dot, grad, div, jump, avg, dS, dS_v, dS_h, action, inner, \
-    outer, sign, cross, CellNormal, lhs, rhs, as_vector, sqrt, Constant, \
-    curl, BrokenElement, FunctionSpace, Projector
+from firedrake import Function, LinearVariationalProblem, LinearVariationalSolver
+from gusto.transport_equation import LinearAdvection
+
+
+def embedded_dg(original_apply):
+    """
+    Decorator to add interpolation and projection steps for embedded
+    DG advection.
+    """
+    def get_apply(self, x_in, x_out):
+        if hasattr(self, "Projector"):
+            def new_apply(self, x_in, x_out):
+                self.xdg_in.interpolate(x_in)
+                original_apply(self, self.xdg_in, self.xdg_out)
+                self.Projector.project()
+                x_out.assign(self.x_projected)
+            return new_apply(self, x_in, x_out)
+
+        else:
+            return original_apply(self, x_in, x_out)
+    return get_apply
 
 
 class Advection(object):
     """
-    Base class for advection schemes for Gusto.
+    Base class for advection schemes.
 
     :arg state: :class:`.State` object.
+    :arg field: field to be advected
+    :arg equation: :class:`.Equation` object, specifying the equation
+    that field satisfies
+    :arg solver_params: solver_parameters
     """
     __metaclass__ = ABCMeta
 
-    def __init__(self, state):
+    def __init__(self, state, field, equation=None, solver_params=None):
+
         self.state = state
-        self.ubar = Function(state.V[0])
+        self.field = field
+
+        if solver_params is None:
+            if isinstance(equation, LinearAdvection):
+                self.solver_parameters = {'ksp_type':'cg',
+                                          'pc_type':'bjacobi',
+                                          'sub_pc_type': 'ilu'}
+            else:
+                self.solver_parameters = {'ksp_type':'preonly',
+                                          'pc_type':'bjacobi',
+                                          'sub_pc_type': 'ilu'}
+        else:
+            self.solver_parameters = solver_params
+        self.dt = self.state.timestepping.dt
+
+        # check to see if we are using an embedded DG method - is we are then
+        # the projector and output function will have been set up in the
+        # equation class and we can get the correct function space from
+        # the output function.
+        if hasattr(equation, "Projector"):
+            fs = equation.xdg_out.function_space()
+            self.Projector = equation.Projector
+            self.xdg_out = equation.xdg_out
+            self.xdg_in = Function(fs)
+            self.x_projected = equation.x_projected
+        else:
+            fs = field.function_space()
```

Comment:
Yes, I agree with your first paragraph. I don't want to move the embedded dg stuff to SSPRK3 because I want to be able to use it with the other advection schemes - I just haven't tested it yet (although see no reason why it shouldn't work). I'd prefer to decorate the abstract apply method but I'm a little wary of doing that without testing all the options.